### ISSUE TITLE
[rush] "tar" exited with code 1 while attempting to create the cache entry

### ISSUE THREAD

**Issue by octogonz** [microsoft/rushstack#2622]
## Summary

I occasionally see the message `"tar" exited with code 1 while attempting to create the cache entry.` printed in yellow during `rush rebuild`.  It is being reported by the cloud build cache (`"buildCache": true` in **experiments.json**).

Yellow indicates a warning. Is this a serious problem?  Am I supposed to do something to fix it?  

At the very least, the log message could be more clear.

 

## Details

I haven't found a way to repro it consistently.

But here's an example log from ddfbd410de775cfa4d81203133bf9042aad3565a

```
Rush Multi-Project Build Tool 5.44.0 - https://rushjs.io
Node.js version is 12.17.0 (LTS)


Starting "rush rebuild"

Executing a maximum of 7 simultaneous processes...

==[ @rushstack/tree-pattern ]====================================[ 1 of 108 ]==
"@rushstack/tree-pattern" completed successfully in 33.82 seconds.

==[ @microsoft/rush-stack-compiler-shared ]======================[ 2 of 108 ]==
"@microsoft/rush-stack-compiler-shared" had an empty script.

==[ @rushstack/eslint-patch ]====================================[ 3 of 108 ]==
"tar" exited with code 1 while attempting to create the cache entry. Rush will attempt to create the cache entry with a JavaScript implementation of tar.

"@rushstack/eslint-patch" completed successfully in 4.20 seconds.

==[ @rushstack/eslint-plugin ]===================================[ 4 of 108 ]==
```

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.44.0 |
| `rushVersion` from rush.json? | 5.44.0 |
| `useWorkspaces` from rush.json? | true |
| Operating system? |  Windows  |
| Node.js version (`node -v`)? | v12.17.0 |

CC @iclanton

**Comment by wbern** [comment]
We've seen this occasionally with the s3 setup, I've seen it while developing locally, personally. Not sure what it means either.

**Comment by iclanton** [comment]
It'd be easy to write a verbose `tar` log somewhere so we can better investigate this. I haven't seen a repro of this personally. It'll be good to get some logs to see what's going on. I'll put that together today.

**Comment by octogonz** [comment]
@iclanton Following up...

Now that verbose logs are enabled, here is what that log shows:

```
Start time: Thu May 13 2021 19:02:45 GMT-0700 (Pacific Daylight Time)
Invoking "C:\WINDOWS\system32\tar.exe -c -f C:\Git\rushstack2\common\temp\build-cache\b86ef9a3ef6e5b6456e496f917543fe1efec20c6 -z -C C:\Git\rushstack2\stack\eslint-patch --files-from C:\Git\rushstack2\stack\eslint-patch\.rush\temp/tarPaths_1620957765802"

======= BEGIN PROCESS OUTPUT =======
[stderr] tar.exe: Failed to clean up compressor
======== END PROCESS OUTPUT ========

Exited with code "1"
```
[This thread](https://forums.freebsd.org/threads/error-tar-failed-to-clean-up-compressor-when-run-from-cron-job-in-freebsd-jail.72573/) say that this error can occur when trying to add the tarball's own filename to the archive.  That doesn't seem like the case here, but maybe it is some other partially written file whose handle was not closed properly. (?)

**Comment by octogonz** [comment]
I cannot try running the commend manually, because `.rush\temp/tarPaths_1620957765802` got deleted.  Looks like this code is deleting it:

https://github.com/microsoft/rushstack/blob/39940938b489665fb77cc1031735e2c3713686e0/apps/rush-lib/src/utilities/TarExecutable.ts#L71

Is it a good idea to delete this file?

**Comment by octogonz** [comment]
Okay I commented out that `deleteFileAsync()` call, then deleted `common\temp\build-cache`, then ran `rush rebuild`.  

It printed the same error about `Failed to clean up compressor`.  Then I went to the folder and manually invoked the `tar.exe -c -f ...` command, and it worked fine.  This supports a theory that `tar` is failing because of some race condition involving file handles.

The inventory of files is very short, though:
```
lib/modern-module-resolution.d.ts
lib/modern-module-resolution.d.ts.map
lib/modern-module-resolution.js
lib/modern-module-resolution.js.map
lib/usage.d.ts
lib/usage.d.ts.map
lib/usage.js
lib/usage.js.map
```

**Comment by octogonz** [comment]
I used Process Monitor to capture a failing run.  Unless I'm mistaken, it seems that `tar.exe` never opens any of these files except for the output file `C:\Git\rushstack2\common\temp\build-cache\fd962512e16c7d12de0581cebc1d73fd04151a57`.

And `node.exe` definitely closes those file handles before `tar.exe` is spawned.

**Comment by octogonz** [comment]
The failure occurs because the `common\temp\build-cache` folder is missing.  I was able to reproduce the above error by invoking `tar` from the command line:  Creating the folder eliminates the error, deleting the folder causes it to reappear.

The error `Failed to clean up compressor` is just a very misleading message. 🤦‍♂️

This should be easy to fix.

**Comment by octogonz** [comment]
For reference, `tar.exe --help` prints this on Windows 10.0.19041:
```
tar(bsdtar): manipulate archive files
First option must be a mode specifier:
  -c Create  -r Add/Replace  -t List  -u Update  -x Extract
Common Options:
  -b #  Use # 512-byte records per I/O block
  -f <filename>  Location of archive (default \\.\tape0)
  -v    Verbose
  -w    Interactive
Create: tar -c [options] [<file> | <dir> | @<archive> | -C <dir> ]
  <file>, <dir>  add these items to archive
  -z, -j, -J, --lzma  Compress archive with gzip/bzip2/xz/lzma
  --format {ustar|pax|cpio|shar}  Select archive format
  --exclude <pattern>  Skip files that match pattern
  -C <dir>  Change to <dir> before processing remaining files
  @<archive>  Add entries from <archive> to output
List: tar -t [options] [<patterns>]
  <patterns>  If specified, list only entries that match
Extract: tar -x [options] [<patterns>]
  <patterns>  If specified, extract only entries that match
  -k    Keep (don't overwrite) existing files
  -m    Don't restore modification times
  -O    Write entries to stdout, don't restore to disk
  -p    Restore permissions (including ACLs, owner, file flags)
bsdtar 3.3.2 - libarchive 3.3.2 zlib/1.2.5.f-ipp
```

**Comment by octogonz** [comment]
Released with Rush 5.47.0
